### ISSUE TITLE
escaping floats with backslashes works but doesn't get saved

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#1277]
I found out you can escape a float with backslashes and you can create a symbol like that

<img width="245" alt="Screen Shot 2021-03-02 at 20 27 15" src="https://user-images.githubusercontent.com/5360142/109728819-b3187480-7b95-11eb-8fe7-9e241faabd8e.png">

the problem is that the backslash disappears from the message box when the patch is reopened

**Comment by umlaeute** [comment]
i think that `\5` is not a valid way to express a symbol `5`.
please don't get in the habit of using it. (languages like *Python* or *C* would take `\5` to mean [&#5;](https://en.wikipedia.org/wiki/Enquiry_character) (aka the *ENQUIRY* character, or `\u0005`).

that said, I agree that not storing the slash in the patch is a bug (like the similar reports #1166, #1083).
while superficially different, i think that all these bugs would be resolved if somebody would sit down and fix the escaping logic for goo.

**Comment by danomatika** [comment]
I agree. I think we need simple, understandable escaping which works within Pd and via `pesend` and `pdrecieve`.

**Comment by Spacechild1** [comment]
> languages like Python or C would take \5 to mean � (aka the ENQUIRY character, or \u0005).

That's a good point. However, Pd's escape sequences work quite differently than in those languages, because the set of special characters is different. For example, Pd doesn't need to escape `'`, `?` or `"`. Nevertheless, I agree that repurposing `\<number>` could be confusing to people coming from those languages.

What about using an unused escape sequence, like `\s`, to force the following characters to be interpreted as a symbol, e.g. `foo \s123`?

Generally, there's a problem with *saving* new escape sequences, because it means that your patch might not work as intended when opening it on older Pd versions...

---

As a side note: Besides Pd's "special" characters (`;`, `,`, `$`), we can only escape whitespace. I think `\n` (newline), `\r` (carriage return) and `\t` (tab) could be easily added to the list, as people coming from C like languages would naturally expect this to work.

In a distant future, we might even support `\uhhh` and `\Uhhh` for typing Unicode symbols :-)

**Comment by danomatika** [comment]
`\s` is a good idea.

**Comment by danomatika** [comment]
I'd recommend closing this issue and opening a new one specifically about improving escape mechanisms.

**Comment by umlaeute** [comment]
i don't know about closing: the original issue described here is that a string `\5` (whether that's valid or not) doesn't get saved as such.
i'm guilty as charged for derailing this into a "`\5` is not a way to turn floats into symbols" discussion (on which i think we all agree by now).

so what should we do with a string `\5`?

**Comment by Spacechild1** [comment]
> so what should we do with a string \5?

Different programming languages/runtimes do it differently. In the case of `foo \s`, Python3 would happily print `foo \s`, while the C/C++ standard libraries would drop the backslash and print `foo s`, but they would also give a compiler warning.

Personally, I would opt for dropping the backslash in invalid escape sequences. Additionally, we could also print an error message, like "unsupported escape sequence ... in symbol ...`.

**Comment by umlaeute** [comment]
my question was more specifically geared towards `\<d>` with `<d>` being an integer.

but anyhow, my libc gives me:

| string in the code | hex representation in the runtime  | notes  |
|------------------------|-----------------------------------------------|-----------------------|
| `\s` | `73 00` | warning: unknown escape sequence: '\s'
| `\5` | `05 00`
| `\25` | `15 00` | octal `025` is *0x15* resp. *21*
| `\22.2` | `12 2E 32 00` | octal `022` is *0x12* resp. *18*
| `\270`  | `B8 00` | octal `0270` is *0xb8* resp. *184*

i probably i have known that `\<num>` (specifically `r"\[0-9]{1,3}"`) is used for the octal representation of a character.

**Comment by danomatika** [comment]
I like this. I think it makes sense to adopt something like C's conventions for this, when possible.

**Comment by porres** [comment]
where is this going? can I close this?
